File: functionapi-api/src/test/kotlin/java/io/github/cottonmc/functionapi/tools/commands/arguments/FunctionAPIFunctionAPIIdentifierArgumentTypeTest.java
package java.io.github.cottonmc.functionapi.tools.commands.arguments;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.tree.LiteralCommandNode;
import io.github.cottonmc.functionapi.api.FunctionAPIIdentifier;
import io.github.cottonmc.functionapi.commands.arguments.FunctionAPIIdentifierArgumentType;
import io.github.cottonmc.functionapi.util.impl.FunctionAPIIdentifierImpl;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;

class FunctionAPIFunctionAPIIdentifierArgumentTypeTest {

    CommandDispatcher<String> stringCommandDispatcher;
    FunctionAPIIdentifier identifier;

    @BeforeEach
    void setUp() throws CommandSyntaxException {
        stringCommandDispatcher = new CommandDispatcher<>();
        LiteralCommandNode<String> register = stringCommandDispatcher.register(
                LiteralArgumentBuilder.<String>literal("id")
                        .then(LiteralArgumentBuilder.<String>literal("id")
                                .executes(context -> {
                                    return 1;
                                }))
                        .then(RequiredArgumentBuilder.<String, FunctionAPIIdentifier>argument("identifier", FunctionAPIIdentifierArgumentType.Companion.identifier())
                                .executes(context -> {
                                    identifier = context.getArgument("identifier", FunctionAPIIdentifierImpl.class);
                                    return 1;
                                }))
        );

        String[] allUsage = stringCommandDispatcher.getAllUsage(stringCommandDispatcher.getRoot(), "", false);
        System.out.println("usages:");
        for (String s : allUsage) {
            System.out.println(s);
        }
        stringCommandDispatcher.execute("id id", "");
    }

    @Test
    @DisplayName("The argument can be read as an functionAPIIdentifier.")
    void canReadArgument() throws CommandSyntaxException {
        stringCommandDispatcher.execute("id namespace:value", "");
        assertAll(
                () -> assertEquals("namespace", identifier.getNamespace()),
                () -> assertEquals("value", identifier.getPath())
        );
    }

    @AfterEach
    void tearDown() {
    }
}

File: functionapi-content/src/test/kotlin/io/github/cottonmc/functionapi/content/FunctionAPIFunctionAPIIdentifierArgumentTypeTest.kt
package io.github.cottonmc.functionapi.content

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.mojang.brigadier.builder.RequiredArgumentBuilder
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.CommandSyntaxException
import io.github.cottonmc.functionapi.api.FunctionAPIIdentifier
import io.github.cottonmc.functionapi.commands.arguments.FunctionAPIIdentifierArgumentType.Companion.identifier
import io.github.cottonmc.functionapi.util.FunctionAPIIdentifierImpl
import org.junit.jupiter.api.*
import org.junit.jupiter.api.function.Executable

internal class FunctionAPIFunctionAPIIdentifierArgumentTypeTest {
    var stringCommandDispatcher: CommandDispatcher<String>? = null
    var identifier: FunctionAPIIdentifier? = null
    @BeforeEach
    @Throws(CommandSyntaxException::class)
    fun setUp() {
        stringCommandDispatcher = CommandDispatcher()
        val register = stringCommandDispatcher!!.register(
                LiteralArgumentBuilder.literal<String>("id")
                        .then(LiteralArgumentBuilder.literal<String>("id")
                                .executes { context: CommandContext<String>? -> 1 })
                        .then(RequiredArgumentBuilder.argument<String, FunctionAPIIdentifier>("identifier", identifier())
                                .executes { context: CommandContext<String> ->
                                    identifier = context.getArgument("identifier", FunctionAPIIdentifierImpl::class.java)
                                    1
                                })
        )
        val allUsage = stringCommandDispatcher!!.getAllUsage(stringCommandDispatcher!!.root, "", false)
        println("usages:")
        for (s in allUsage) {
            println(s)
        }
        stringCommandDispatcher!!.execute("id id", "")
    }

    @Test
    @DisplayName("The argument can be read as an functionAPIIdentifier.")
    @Throws(CommandSyntaxException::class)
    fun canReadArgument() {
        stringCommandDispatcher!!.execute("id namespace:value", "")
        Assertions.assertAll(
                Executable { Assertions.assertEquals("namespace", identifier!!.namespace) },
                Executable { Assertions.assertEquals("value", identifier!!.path) }
        )
    }

    @AfterEach
    fun tearDown() {
    }
}